src/pipx/commands/upgrade.py

```diff
@@ -30,18 +31,22 @@ def upgrade(
 
     venv = Venv(venv_dir, verbose=verbose)
 
+    if not venv.package_metadata:
+        print(
+            f"Not upgrading {red(bold(package))}.  It has missing internal pipx metadata.\n"
+            f"    It was likely installed using a pipx version before 0.15.0.0.\n"
+            f"    Please uninstall and install this package, or reinstall-all to fix."
+        )
+        return 0
+
     package_metadata = venv.package_metadata[package]
-    if package_metadata.package_or_url is not None:
-        package_or_url = package_metadata.package_or_url
-        old_version = package_metadata.package_version
-        include_apps = package_metadata.include_apps
-        include_dependencies = package_metadata.include_dependencies
-    else:
-        # fallback if no metadata
-        package_or_url = package
-        old_version = ""
-        include_apps = True
-        include_dependencies = False
+    if package_metadata.package_or_url is None:
```

Comment:
Note: this should never happen, but type-checking catches this because `package_or_url` is type `Optional[str]`.  In `pipx_metadata_file.py` we specifically check that `package_or_url` is not `None` in `PipxMetadata._validate_before_write()` before writing metadata.